server/telemetry/telemetry_client.go

```diff
@@ -0,0 +1,230 @@
+)
+
+var (
+	telemetryInterval = flag.Duration("telemetry_interval", 24*time.Hour, "How often telemetry data will be reported")
+	verboseTelemetry  = flag.Bool("verbose_telemetry_client", false, "If true; print telemetry client information")
+	disableTelemetry  = flag.Bool("disable_telemetry", false, "If true; telemetry will be disabled")
+	telemetryEndpoint = flag.String("telemetry_endpoint", "grpcs://t.buildbuddy.io", "The telemetry endpoint to use")
+)
+
+type TelemetryClient struct {
+	env    environment.Env
+	ticker *time.Ticker
+	quit   chan struct{}
+
+	version          string
+	instanceUUID     string
+	installationUUID string
+
+	failedLogs []*telpb.TelemetryLog
+}
+
+func NewTelemetryClient(env environment.Env) *TelemetryClient {
+	return &TelemetryClient{
+		env:              env,
+		version:          getAppVersion(),
+		instanceUUID:     getInstanceUUID(),
+		installationUUID: getInstallationUUID(env),
+		failedLogs:       []*telpb.TelemetryLog{},
+	}
+}
+
+func (t *TelemetryClient) Start() {
+	t.ticker = time.NewTicker(*telemetryInterval)
+	t.quit = make(chan struct{})
+
+	if *disableTelemetry {
+		printIfVerbose("Telemetry disabled")
+		return
+	}
+
+	go func() {
+		for {
+			select {
+			case <-t.ticker.C:
+				t.logTelemetryData()
+			case <-t.quit:
+				printIfVerbose("Telemetry task %d exiting.", 0)
+				break
+			}
+		}
+	}()
+
+	// Log once on startup.
+	t.logTelemetryData()
+}
+
+func (t *TelemetryClient) Stop() {
+	close(t.quit)
+	t.ticker.Stop()
+}
+
+func (t *TelemetryClient) logTelemetryData() {
+	ctx := context.Background()
+	conn, err := grpc_client.DialTarget(*telemetryEndpoint)
+	client := telpb.NewTelemetryClient(conn)
+
+	log := &telpb.TelemetryLog{
+		InstallationUuid: t.installationUUID,
+		InstanceUuid:     t.instanceUUID,
+		LogUuid:          getLogUUID(),
+		RecordedAtUsec:   time.Now().UnixNano() / int64(time.Microsecond),
+		AppVersion:       t.version,
+		AppUrl:           t.env.GetConfigurator().GetAppBuildBuddyURL(),
+		Hostname:         getHostname(),
+		DailyStat:        &telpb.TelemetryStat{},
+		Feature:          getFeatures(t.env),
+	}
+
+	// Fill invocation related stats
+	if err := t.env.GetInvocationDB().FillCounts(ctx, log.DailyStat); err != nil {
+		printIfVerbose("Error getting telemetry invocation counts: %s", err)
+	}
+
+	// Fill user related stats
+	if err := t.env.GetUserDB().FillCounts(ctx, log.DailyStat); err != nil {
+		printIfVerbose("Error getting telemetry invocation counts: %s", err)
+	}
+
+	req := &telpb.LogTelemetryRequest{
+		Log: append(t.failedLogs, log),
+	}
+
+	response, err := client.LogTelemetry(ctx, req)
+	if err != nil || response.Status.Code != 0 {
+		printIfVerbose("Error posting telemetry data: %s", err)
+		t.failedLogs = append(t.failedLogs, log)
+		return
+	}
+
+	t.failedLogs = []*telpb.TelemetryLog{}
+	printIfVerbose("Telemetry data posted: %+v", response)
+}
+
+func printIfVerbose(message string, args ...interface{}) {
```

Comment:
TODO(tylerw): use glog or similar. Then we can just do log.Verbose(blah) here.

Reply:
Done